File: app/src/main/java/rbsoftware/friendstagram/SelectImageFragment.java
package rbsoftware.friendstagram;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.FileProvider;
import android.support.v4.content.Loader;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.io.File;

import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.OnPermissionDenied;
import permissions.dispatcher.OnShowRationale;
import permissions.dispatcher.PermissionRequest;
import permissions.dispatcher.RuntimePermissions;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link ImageSelectListener} interface
 * to handle interaction events.
 * Use the {@link SelectImageFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
@RuntimePermissions
public class SelectImageFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor>, ImageSelectListener {

    private static final int EXTERNAL_STORAGE_ID = 0;
    private static final int INTERNAL_STORAGE_ID = 1;
    private static final int REQUEST_IMAGE_CAPTURE = 2;
    private static final String TAG = "SelectImageFragment";

    private RecyclerView recyclerView;
    private PicturesAdapter adapter;

    private ImageSelectListener mListener;

    public SelectImageFragment() {
        // Required empty public constructor
    }

    public static SelectImageFragment newInstance() {
        return new SelectImageFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_select_image, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        recyclerView = (RecyclerView) view.findViewById(R.id.rv);
        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), 3));

        if (adapter != null) {
            recyclerView.setAdapter(adapter);
        }


        FloatingActionButton fab = (FloatingActionButton) view.findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SelectImageFragmentPermissionsDispatcher.takePictureWithCheck(SelectImageFragment.this);
            }
        });
        SelectImageFragmentPermissionsDispatcher.loadImagesWithCheck(this);
    }

    @TargetApi(Build.VERSION_CODES.M)
    @NeedsPermission(Manifest.permission.READ_EXTERNAL_STORAGE)
    void loadImages() {
        getLoaderManager().initLoader(EXTERNAL_STORAGE_ID, null, this);
        // TODO: See if any phones use the internal storage ID for photos
//        getLoaderManager().restartLoader(INTERNAL_STORAGE_ID, null, this);
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnShowRationale(Manifest.permission.READ_EXTERNAL_STORAGE)
    void showRationaleForLoadingImages(final PermissionRequest request) {
        new AlertDialog.Builder(getContext())
                .setMessage(R.string.permission_read_storage_rationale)
                .setPositiveButton(R.string.btn_allow, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        request.proceed();
                    }
                })
                .setNegativeButton(R.string.btn_deny, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        request.cancel();
                    }
                })
                .show();
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnPermissionDenied(Manifest.permission.READ_EXTERNAL_STORAGE)
    void showDeniedForLoadingImages() {
        Toast.makeText(getContext(), R.string.permission_read_storage_denied, Toast.LENGTH_SHORT).show();
    }

    @TargetApi(Build.VERSION_CODES.M)
    @NeedsPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    void takePicture() {
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if (takePictureIntent.resolveActivity(getActivity().getPackageManager()) != null) {
            File photo = ImageHandler.createImageFile();

            if (photo != null) {
                Log.d(TAG, "takePicture: " + photo.toString());
                Uri photoURI = FileProvider.getUriForFile(getContext(), "rbsoftware.friendstagram.fileprovider", photo);
                Log.d(TAG, "takePicture: " + photoURI);
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
                startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
                ImageHandler.addToMediaScanner(photoURI, getContext());
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnShowRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    void showRationaleForSavingImages(final PermissionRequest request) {
        new AlertDialog.Builder(getContext())
                .setMessage(R.string.permission_write_storage_rationale)
                .setPositiveButton(R.string.btn_allow, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        request.proceed();
                    }
                })
                .setNegativeButton(R.string.btn_deny, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        request.cancel();
                    }
                })
                .show();
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnPermissionDenied(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    void showDeniedForSavingImages() {
        Toast.makeText(getContext(), R.string.permission_write_storage_denied, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        // NOTE: delegate the permission handling to generated method
        SelectImageFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof ImageSelectListener) {
            mListener = (ImageSelectListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement ImageSelectListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        Log.d(TAG, "onCreateLoader: Loader ID = " + Integer.toString(id));
        return new CursorLoader(
                getContext(),
                id == EXTERNAL_STORAGE_ID ?
                        MediaStore.Images.Media.EXTERNAL_CONTENT_URI :
                        MediaStore.Images.Media.INTERNAL_CONTENT_URI,
                new String[]{
                        MediaStore.Images.ImageColumns._ID,
                        MediaStore.Images.ImageColumns.DATA
                },
                null,
                null,
                MediaStore.Images.ImageColumns._ID + " DESC"
        );
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        Log.d(TAG, "onLoadFinished: cursor: " + (data == null ? "null" : Integer.toString(data.getCount()) + " elements"));
        if (adapter == null) {
            recyclerView.setAdapter(adapter = new PicturesAdapter(data, getContext(), this));
        } else {
            adapter.changeCursor(data);
        }

    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
        if (adapter != null) {
            Log.i(TAG, "onLoaderReset: Resetting loader");
            adapter.changeCursor(null);
        }
    }

    private void updateMediaScanner(Uri photoURI) {
        Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        mediaScanIntent.setData(photoURI);
        getActivity().sendBroadcast(mediaScanIntent);
    }

    @Override
    public void onImageSelected(Uri url) {
        mListener.onImageSelected(url);
    }
}

File: app/src/main/java/rbsoftware/friendstagram/ui/fragment/SelectImageFragment.kt
package rbsoftware.friendstagram.ui.fragment

import android.Manifest
import android.annotation.TargetApi
import android.content.Intent
import android.database.Cursor
import android.os.Build
import android.os.Bundle
import android.provider.MediaStore
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.support.v4.app.LoaderManager
import android.support.v4.content.CursorLoader
import android.support.v4.content.FileProvider
import android.support.v4.content.Loader
import android.support.v7.app.AlertDialog
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import permissions.dispatcher.*
import rbsoftware.friendstagram.ImageHandler
import rbsoftware.friendstagram.ImageSelectListener
import rbsoftware.friendstagram.R
import rbsoftware.friendstagram.ui.adapter.PicturesAdapter

/**
 * Created by Rushil on 8/18/2017.
 */
@RuntimePermissions()
class SelectImageFragment : Fragment(), LoaderManager.LoaderCallbacks<Cursor> {
    private var adapter: PicturesAdapter? = null
    private var recyclerView: RecyclerView? = null
    private lateinit var onImageSelected: ImageSelectListener

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater?.inflate(R.layout.fragment_select_image, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView = view?.findViewById(R.id.rv)
        val fab: FloatingActionButton? = view?.findViewById(R.id.fab)

        recyclerView?.layoutManager = GridLayoutManager(context, 3)
        adapter?.let { recyclerView?.adapter = it }

        fab?.setOnClickListener { SelectImageFragmentPermissionsDispatcher.takePictureWithCheck(this) }
        SelectImageFragmentPermissionsDispatcher.loadImagesWithCheck(this)
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor>? {
        Log.d(TAG, "Loader created")
        return CursorLoader(
                context,
                if (id == EXTERNAL_STORAGE_ID)
                    MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                else
                    MediaStore.Images.Media.INTERNAL_CONTENT_URI,
                arrayOf(
                        MediaStore.Images.ImageColumns._ID,
                        MediaStore.Images.ImageColumns.DATA
                ),
                null,
                null,
                MediaStore.Images.ImageColumns._ID + " DESC"
        )
    }

    override fun onLoadFinished(loader: Loader<Cursor>?, data: Cursor?) {
        Log.d(TAG, "Finished loading")
        if (adapter == null) {
            adapter = data?.let { PicturesAdapter(it, context) }
            adapter?.setOnImageSelectListener(onImageSelected)
            recyclerView?.adapter = adapter
        } else {
            adapter?.changeCursor(data)
        }
    }

    override fun onLoaderReset(loader: Loader<Cursor>?) {
        adapter?.changeCursor(null)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        // NOTE: delegate the permission handling to generated method
        SelectImageFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults)
    }

    fun setOnImageSelectListener(imageSelectListener: ImageSelectListener) {
        this.onImageSelected = imageSelectListener
        adapter?.setOnImageSelectListener(onImageSelected)
    }

    @TargetApi(Build.VERSION_CODES.M)
    @NeedsPermission(Manifest.permission.READ_EXTERNAL_STORAGE)
    fun loadImages() {
        loaderManager.initLoader(EXTERNAL_STORAGE_ID, null, this)
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnShowRationale(Manifest.permission.READ_EXTERNAL_STORAGE)
    internal fun showRationaleForLoadingImages(request: PermissionRequest) {
        AlertDialog.Builder(context)
                .setMessage(R.string.permission_read_storage_rationale)
                .setPositiveButton(R.string.btn_allow) { _, _ -> request.proceed() }
                .setNegativeButton(R.string.btn_deny) { _, _ -> request.cancel() }
                .show()
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnPermissionDenied(Manifest.permission.READ_EXTERNAL_STORAGE)
    internal fun showDeniedForLoadingImages() {
        Toast.makeText(context, R.string.permission_read_storage_denied, Toast.LENGTH_SHORT).show()
    }

    @TargetApi(Build.VERSION_CODES.M)
    @NeedsPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    fun takePicture() {
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (takePictureIntent.resolveActivity(activity.packageManager) != null) {
            val photo = ImageHandler.createImageFile()

            Log.d(TAG, "takePicture: " + photo.toString())
            val photoURI = FileProvider.getUriForFile(context, "rbsoftware.friendstagram.fileprovider", photo)
            Log.d(TAG, "takePicture: " + photoURI)
            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
            startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
            ImageHandler.addToMediaScanner(photoURI, context)
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnShowRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    internal fun showRationaleForSavingImages(request: PermissionRequest) {
        AlertDialog.Builder(context)
                .setMessage(R.string.permission_write_storage_rationale)
                .setPositiveButton(R.string.btn_allow) { _, _ -> request.proceed() }
                .setNegativeButton(R.string.btn_deny) { _, _ -> request.cancel() }
                .show()
    }

    @TargetApi(Build.VERSION_CODES.M)
    @OnPermissionDenied(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    internal fun showDeniedForSavingImages() {
        Toast.makeText(context, R.string.permission_write_storage_denied, Toast.LENGTH_SHORT).show()
    }

    companion object {
        private const val EXTERNAL_STORAGE_ID = 0
        private const val INTERNAL_STORAGE_ID = 1
        private const val REQUEST_IMAGE_CAPTURE = 2
        private const val TAG = "SelectImageFragment"

        fun newInstance(): SelectImageFragment {
            return SelectImageFragment()
        }
    }
}